lib/recog/nizer.rb

```diff
@@ -20,43 +20,122 @@ class Nizer
   }
 
   @@db_manager = nil
+  @@db_sorted = false
 
+
+  #
+  # Load fingerprints from a specific file or directory
+  # This will not preserve any fingerprints that have already been loaded
+  # @param path [String] Path to file or directory of XML fingerprints
+  def self.load_db(path = nil)
+    if path
+      @@db_manager = Recog::DBManager.new(path)
+    else
+      @@db_manager = Recog::DBManager.new
+    end
+
+    # Sort the databases, no behavior or result change for those calling
+    # Nizer.match or Nizer.multi_match as they have a single DB
+    @@db_manager.databases.sort! { |a, b| b.preference <=> a.preference }
+    @@db_sorted = true
+  end
+
+  #
+  # Destroy the current DBManager object
+  def self.unload_db
+    @@db_manager = nil
+    @@db_sorted = false
+  end
+
+  #
+  # Display the fingerprint databases in the order in which they will be used
+  # to match banners.  This is useful for fingerprint tuning and debugging.
+  def self.display_db_order
+    self.load_db unless @@db_manager
+
+    puts format('%s  %-22s  %s ', 'Preference', 'Database', 'Type')
+    @@db_manager.databases.each do |db|
+      puts format('%10.3f  %-22s  %s ', db.preference, db.match_key,
+                  db.database_type)
+    end
+  end
+
+  #
+  # 2016.11 - Rewritten to be wrapper around #match_db_all, functionality
+  # and results must remain unchanged.
   #
   # Locate a database that corresponds with the `match_key` and attempt to
-  # find a matching {Fingerprint fingerprint}, stopping at the first hit. Returns `nil`
-  # when no matching database or fingerprint is found.
+  # find a matching {Fingerprint fingerprint}, stopping at the first hit.
+  # Returns `nil` when no matching database or fingerprint is found.
   #
   # @param match_key [String] Fingerprint DB name, e.g. 'smb.native_os'
-  # @return (see Fingerprint#match)
+  # @param match_string [String] String to match
+  # @return (see Fingerprint#match) or nil
   def self.match(match_key, match_string)
-    match_string = match_string.to_s.unpack("C*").pack("C*")
-    @@db_manager ||= Recog::DBManager.new
-    @@db_manager.databases.each do |db|
-      next unless db.match_key == match_key
-      db.fingerprints.each do |fprint|
-        m = fprint.match(match_string)
-        return m if m
-      end
-    end
+    filter = { match_key: match_key, multi_match: false }
+    matches = self.match_all_db(match_string, filter)
+
+    return matches[0] if matches && matches[0]
```

Comment:
matches will be empty, not `nil`, according to code and docs.  You replace this and the next line with simply `return matches[0]`.